### ISSUE TITLE
EPIC: LLM-Style Capabilities with AST-First Architecture

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#696]
## Goal
Extend Klareco to support LLM-style capabilities (text generation, instruction following, chat, reasoning) while maintaining the AST-first architecture's advantages: grammatical correctness, explainability, and efficiency.

## Motivation
Traditional LLMs achieve these capabilities through massive pretraining (70B+ params) and learn everything from text. Klareco's thesis: By making grammar 100% deterministic via ASTs, we can achieve comparable capabilities with **<100M learned parameters**, focusing learned capacity entirely on reasoning.

## Architecture Comparison

| Capability | Traditional LLM | Klareco AST-First |
|------------|-----------------|-------------------|
| **Text generation** | Next-token prediction | Next-AST-node prediction (#692) |
| **Instruction following** | Fine-tune on text pairs | Parse + intent routing (#693) |
| **Chat** | Context window (text) | Discourse model (ASTs) (#694) |
| **Reasoning** | Emergent from pretraining | Symbolic + minimal neural (#695) |
| **Grammar correctness** | Learned (unreliable) | Deterministic (100%) |
| **Explainability** | None | AST Trail (full provenance) |
| **Parameters** | 70B+ | <100M |

### 1. Text Generation (#692)
**Next-AST-Node Prediction (~20M params)**
- Predict next AST node instead of text token
- Ensures 100% grammatical validity by construction
- Used for: completion, paraphrasing, question generation

### 2. Instruction Following (#693)
**Intent Classification + Routing (~5M params)**
- Parse instruction to AST
- Classify intent (QA, summarization, translation, generation)
- Route to appropriate expert module
- Deterministic patterns + learned classifier

### 3. Multi-Turn Chat (#694)
**Discourse Model + Coreference (~10M params)**
- Track conversation as AST forest
- Deterministic coreference resolution
- Learned discourse relation prediction
- Entity tracking across turns

### 4. Symbolic Reasoning (#695)
**Inference Engine (0 params - deterministic!)**
- Extract facts from retrieved ASTs
- Apply first-order logic rules
- Multi-hop reasoning with full explainability
- AST Trail for provenance tracking

## Training Pipeline

```
Stage 0: Parser (deterministic, 0 params)
  ↓
Stage 1: Root Embeddings (~500K params) [IN PROGRESS]
  ↓
Stage 2: Compositional Embeddings (~500K params) [IN PROGRESS]
  ↓
Stage 3: Intent Classifier (~5M params) [#693]
  ↓
Stage 4: AST Generator (~20M params) [#692]
  ↓
Stage 5: Discourse Model (~10M params) [#694]
  ↓
Stage 6: Reasoning (deterministic) [#695]
  ↓
Total Learned Parameters: ~36M (vs 70B+ for traditional LLMs!)
```

### Phase 1: Foundation (Current)
- [x] Parser (16 Esperanto rules)
- [x] TreeLSTMEncoder
- [x] Two-stage retrieval
- [x] Extractive QA
- [ ] Root embeddings training (#685)
- [ ] Compositional embeddings

### Phase 2: Instruction Following (Q2 2026)
- [ ] Intent classifier (#693)
- [ ] Instruction handler
- [ ] CLI: `klareco instruct`

### Phase 3: Generation (Q2-Q3 2026)
- [ ] AST generator (#692)
- [ ] Training data preparation
- [ ] CLI: `klareco generate`

### Phase 4: Reasoning (Q3 2026)
- [ ] Fact extraction enhancement (#695)
- [ ] Inference engine (deterministic)
- [ ] AST Trail system
- [ ] Multi-hop QA evaluation

### Phase 5: Chat (Q4 2026)
- [ ] Coreference tracker (#694)
- [ ] Discourse model
- [ ] Conversation state management
- [ ] CLI: `klareco chat --interactive`

### Functionality
- [ ] Generate grammatically correct Esperanto text (100% parse rate)
- [ ] Follow instructions across 5+ task types (QA, summarization, etc.)
- [ ] Maintain coherent 10+ turn conversations
- [ ] Perform multi-hop reasoning with explainability

### Performance
- [ ] Generation quality: Human eval >80% coherent
- [ ] Instruction accuracy: >95% correct routing
- [ ] Chat coherence: >85% human evaluation
- [ ] Reasoning accuracy: >70% on multi-hop QA

### Efficiency
- [ ] Total learned parameters: <100M
- [ ] Inference speed: <500ms per query
- [ ] Memory footprint: <2GB RAM

### Explainability
- [ ] Every answer includes AST Trail
- [ ] Users can inspect reasoning steps
- [ ] Grammar errors: 0% (guaranteed by AST structure)

### Create Esperanto Benchmarks
1. **Instruction Following**: 500 instructions across task types
2. **Multi-Turn Chat**: 100 dialogues, 5-10 turns each
3. **Multi-Hop Reasoning**: 500 questions requiring 2-5 inference steps
4. **Text Generation**: 200 completion/generation tasks

### Comparison Baselines
- GPT-3.5/GPT-4 (via translation)
- LLaMA-2-Chat (via translation)
- Esperanto-specific LLMs (if available)

## Key Advantages of AST-First Approach

1. **Grammatical Correctness**: Impossible to generate invalid Esperanto
2. **Explainability**: Full provenance via AST Trail
3. **Efficiency**: 100-1000x fewer parameters than traditional LLMs
4. **Data Efficiency**: Less training data needed (grammar is programmatic)
5. **Debugging**: Can inspect ASTs at every step
6. **Controllability**: Can modify reasoning rules without retraining

## Related Issues
- #685: Root embeddings (foundation)
- #686: Reranker training
- #687: M1 selectional preference
- #680, #690: Answer generation
- #654: Semantic ontology
- #692: AST generation (**NEW**)
- #693: Instruction following (**NEW**)
- #694: Multi-turn chat (**NEW**)
- #695: Symbolic reasoning (**NEW**)

## Documentation Needed
- [ ] Architecture diagram: AST-first vs traditional LLM
- [ ] Training guide for each stage
- [ ] Evaluation methodology
- [ ] User guide: CLI commands for each capability
- [ ] Developer guide: Extending with new capabilities

## Future Vision
Once these capabilities are proven, Klareco demonstrates that:
- **Grammar doesn't need to be learned** (save parameters for reasoning)
- **Reasoning can be explicit** (symbolic rules + minimal neural)
- **Small models can match large models** (on structured tasks)
- **Explainability is achievable** (AST Trail provides full provenance)

This validates the core thesis: **Pure Esperanto AI with AST-first architecture achieves LLM capabilities at 100x smaller scale**.

**Comment by marctjones** [comment]
## Architectural Analysis Complete

See #697 and `docs/DETERMINISTIC_VS_LEARNED.md` for detailed breakdown of deterministic vs learned components.

### Key Decision Points

**Minimal Configuration (21M params)**:
- Foundation: 1M
- Text Generation: 20M (required)
- Instruction Following: 0M (patterns only)
- Chat: 0M (rules only)
- Reasoning: 0M (rules only)

**Full Configuration (36M params)**:
- Adds learned intent classifier: +5M
- Adds learned discourse model: +10M

Both are **1,944x - 4,861x smaller** than traditional LLMs (70B-175B params).

### Recommendation
Start with **Minimal Configuration** to prove the thesis that deterministic-first design can achieve LLM capabilities with <25M params. Add optional learned components in Phase 2 if needed for quality improvements.

### Deterministic Coverage
- **Instruction Following**: 85% deterministic (verb patterns)
- **Multi-Turn Chat**: 90% deterministic (Esperanto pronouns!)
- **Symbolic Reasoning**: 100% deterministic (logic rules)
- **Text Generation**: 30% deterministic (parsing/deparsing)

This validates the core thesis: Grammar doesn't need to be learned.

**Comment by marctjones** [comment]
## 🎯 Model Naming Update

All models now have **meaningful, purpose-driven names** instead of generic labels (M0.1, M1, etc.)

### New Model Names

| Old Label | New Name | Purpose |
|-----------|----------|---------|
| M0.1 | **RootEmbedder** | Embeds roots semantically |
| M0.2 | **MorphemeComposer** | Composes morphemes into words |
| M1 | **PlausibilityFilter** | Filters implausible facts |
| M2 | **RelevanceRanker** | Ranks by relevance |
| M3 | **ASTEncoder** | Encodes AST to vector |
| M4 | **NodePredictor** | Predicts next AST node |
| M5 | **IntentClassifier** | Classifies instruction intent |
| M6 | **DiscourseClassifier** | Classifies discourse relations |

### Benefits

✅ **Self-documenting**: "PlausibilityFilter" is clearer than "M1"
✅ **Easier communication**: "Train the RootEmbedder" vs "Train M0.1"
✅ **Professional naming**: Follows industry standards
✅ **Memorable**: Purpose-based names stick

### Updated Documentation

- ✅ `docs/MODEL_NAMING.md` - Complete naming specification
- ✅ `docs/MODEL_INVENTORY.md` - Updated catalog
- ✅ `docs/MODEL_ISSUES_REFERENCE.md` - Updated issue mapping
- ✅ `docs/DETERMINISTIC_VS_LEARNED.md` - Updated analysis
- ✅ `docs/MODEL_NAMES_SUMMARY.md` - Quick reference

### Configurations with New Names

**Minimal (21M)**: RootEmbedder, MorphemeComposer, ASTEncoder, NodePredictor
**Standard (28M)**: + PlausibilityFilter, RelevanceRanker
**Full (43M)**: + IntentClassifier, DiscourseClassifier

See `docs/MODEL_NAMING.md` for complete specification.
